colorama/ansitowin32.py

```diff
@@ -18,7 +18,7 @@ def is_stream_closed(stream):
 
 
 def is_a_tty(stream):
-    if 'PYCHARM_HOSTED' in os.environ: 
+    if 'PYCHARM_HOSTED' in os.environ and (stream is sys.__stdout__ or stream is sys.__stderr__):
```

Comment:
Do we want to check against the `__` versions of `stdout` and `stderr` instead of using `sys.stdout` and `sys.stderr`? If there was another redirection (so the values of the "regular" sys streams changed), we would still want to return "True" under PyCharm for these streams, no? Since `init` uses `sys.stdout` and `sys.stderr` directly, it makes sense to check against these streams. Perhaps I'm missing something though.

Reply:
My reasoning was the following: the `stdout` and `stderr` under Pycharm are inherently bugged. These two streams are unambiguously accessible through `__stdout__` and `__stderr__`. If the user happens to override `sys.stdout`, then we do not have much information about the stream that will replace it (it could be a `file` or a `StringIO`), so we should not assume that the replacing stream is bugged too, instead we should check as usual using `stream.isatty()`.

Basically, the edge case is an user replacing `sys.stdout` with a non-tty stream on Pycharm. In such a case, `colorama.is_a_tty(sys.stdout)` should return `False`, but this would not be the case if the check is made using `stream is sys.stdout`.

Does it make sense to you? I do not exclude having missed something too.

Reply:
I think I understand (a user custom `sys.stdout` may not be the real "PyCharm terminal" and so may need to be treated as "not a tty"). The problem is that after "colorama.init", `sys.stdout` will be replaced by colorama and then `colorama.is_a_tty(sys.stdout)` will return False (in PyCharm) although it is a simple wrapper of the "original". Maybe we should check for both? i.e. something like
`... and (stream in (sys.__stdout__, sys.__stderr__, sys.stdout, sys.stderr))`?

If I understand correctly, the problem is generally only when using the PyCharm interactive console, and shouldn't affect scripts running outside of the IDE.

Reply:
Oh, you are right, I did not even think of it, thanks for pointing this out...

But I do not see how checking for `stream in ...` will prevent the edge case of user replacing `sys.stderr` with a non-tty stream (on Pycharm). 😕 

In my opinion, as we are wrapping a broken `sys.__stderr__` stream, we should take the opportunity to fix it. Something like this:

```python
class StreamWrapper(object):
    ...

    def isatty(self):
        return is_a_tty(self.__wrapped)
```

It seems to fix all the edge cases we discussed, what do you think?